x-pack/plugins/infra/server/lib/alerting/log_threshold/log_threshold_executor.ts

```diff
@@ -985,3 +1073,91 @@ export const FIRED_ACTIONS: ActionGroup<'logs.threshold.fired'> = {
     defaultMessage: 'Fired',
   }),
 };
+
+const getContextAggregation = (
+  params: Pick<RuleParams, 'timeSize' | 'timeUnit' | 'groupBy'> & { criteria: CountCriteria }
+) => {
+  const validPrefixForContext = ['host', 'cloud', 'orchestrator', 'container', 'labels', 'tags'];
+  const positiveCriteria = params.criteria.filter((criterion: Criterion) =>
+    positiveComparators.includes(criterion.comparator)
+  );
+
+  const fieldsFromGroupBy = params.groupBy
+    ? getFieldsSet(params.groupBy, validPrefixForContext)
+    : new Set<string>();
+  const fieldsFromCriteria = getFieldsSet(
+    positiveCriteria.map((criterion: Criterion) => criterion.field),
+    validPrefixForContext
+  );
+  const fieldsPrefixList = Array.from(
+    new Set<string>([...fieldsFromGroupBy, ...fieldsFromCriteria])
+  );
+  const fieldsList = fieldsPrefixList.map((prefix) => (prefix === 'tags' ? prefix : `${prefix}.*`));
+
+  const additionalContextAgg =
+    fieldsList.length > 0
+      ? {
+          additionalContext: {
+            top_hits: {
+              size: 1,
+              fields: fieldsList,
+              _source: false,
+            },
+          },
+        }
+      : null;
+
+  return additionalContextAgg;
+};
+
+const getAvailableFields = async (
+  esClient: ElasticsearchClient,
+  index: string
+): Promise<Record<string, Record<string, FieldCapsFieldCapability>>> => {
+  // Get all supported fields
+  const respMapping = await esClient.fieldCaps({
+    index,
+    fields: '*',
+  });
+
+  return respMapping.fields;
+};
+
+const getFieldsSet = (groupBy: string[] | undefined, validPrefix: string[]): Set<string> => {
+  return new Set<string>(
+    groupBy
+      ?.map((currentGroupBy) => currentGroupBy.split('.')[0])
+      .filter((groupByPrefix) => validPrefix.includes(groupByPrefix))
+  );
+};
+
+const hostFieldsToExclude = ['disk', 'network', 'cpu'];
+
+const formatFields = (
```

Comment:
Can you help me understand why we need such complex formatting logic here? I'd like to understand what the criteria for extracting the first array element are. For example, a host usually has multiple MAC addresses, but this just picks the first. At the same time it preserves multiple IP addresses.

Has something similar been done for other rule types?

Reply:
Maybe this helps -
```
if ((fieldType === ES_FIELD_TYPES.KEYWORD ||
      fieldType === ES_FIELD_TYPES.TEXT ||
      fieldType === ES_FIELD_TYPES.OBJECT) &&
    value.length === 1
  ) {
    formattedContextFields[key] = value?.[0];
  } else {
    formattedContextFields[key] = value;
}
```
Added check to see if array length is one. If there are multiple MAC addresses, it will remain as an array.

Reply:
I don't know if we want to burden ourselves with having to maintain a comprehensive list of all the ECS fields that we want to treat as arrays. :thinking: What would be the downside of just keeping all the arrays? Then the user can choose to access the first or all of the values.

Reply:
The part of reason I implemented this way was to match with format of additional context in `infra` rules where we are using `_source`. If we keep arrays, there will be different format for same functionality in `infra` rules and `log threshold rule`. For example, you can use `host.name` in `infra` rules, whereas in `log threshold rule`, you would need to use `host.name[0]`.

Second reason is that it's not possible to write the context fields with arrays to AAD because it doesn't match with index mapping. Here `host.name` is an array, but AAD index expects it to be `keyword`, so we need to somehow map it to correct types that AAD index expects.

Reply:
Thanks for the explanation!

---

> [...] to match with format of additional context in `infra` rules where we are using `_source`.

Yes, being consistent is definitely important for the user experience. What is the plan for the other rules that currently rely on `_source`? Will that change in the future? Will they ever support runtime fields? Or will they rely on synthetic source?

---

> Second reason is that it's not possible to write the context fields with arrays to AAD because it doesn't match with index mapping. Here `host.name` is an array, but AAD index expects it to be `keyword`, so we need to somehow map it to correct types that AAD index expects.

Right, I forgot that this gets written to an index too and not just used in the message template.

But still, any field mapped as `keyword` can still contain a list of terms. So wouldn't any client retrieving data from AAD indices have to handle the fact that there could be more than one value for any field anyway? If so, maybe the knowledge of how to interpret each field is already maintained somewhere?

---

Coming back to the cost/benefit question: If all other observability alerting rules require `_source`, maybe we value simplicity and consistency higher than supporting runtime fields in log threshold alerts? That sounds like a product question. :thinking:

Reply:
The ecs fields specify whether it contains array or not. While fields mapped as `keyword` can contain list of terms, it is still defined whether it will contain list or not. If we try to index array when it doesn't support array, it throws an error. So if we want to index it, we have to map it to correct type and remove array when it is not supported.

https://github.com/elastic/kibana/blob/7693b2fbee404dc16a5d9514226d82f722f6ec16/x-pack/plugins/rule_registry/common/assets/field_maps/ecs_field_map.ts#L1584

```
  'host.name': {
    type: 'keyword',
    array: false,
    required: false,
  },
  'host.mac': {
    type: 'keyword',
    array: true,
    required: false,
  },
```

Another thing to note, while we can add runtime fields in context and use them in message template, I am not sure if we can index them in AAD. This check is already there in code to see if field is present in the mapping.

I don't know if there is plan to use `fields` in other rules. @katrin-freihofner @grabowskit @simianhacker may have more information.

Reply:
We could also check types against `ecs_field_map.ts` instead of with `fieldCaps` API. That way we can be sure that no arrays are omitted. With `fieldCaps` API, we don't get information whether it's an array.

Reply:
> Another thing to note, while we can add runtime fields in context and use them in message template, I am not sure if we can index them to AAD.

Sorry, I should have been more careful to avoid the confusion about where runtime fields are supported. It's not about the mapping of the AAD indices, but the fields in the user's log entries that we read values like `host.name` from. In the case of logs it could be that the value for `host.name` is extracted from the log message at query time using a runtime field. When storing it in the AAD indices it doesn't matter whether the value was computed at runtime or not.

> We could also check types against `ecs_field_map.ts` instead of with `fieldCaps` API.

Yes, that sounds like a good source of truth :+1:

Reply:
Ah, interesting! Thanks for the explanation.

I will make changes to use mapping from `ecs_field_map.ts`.

Reply:
Done the changes.